File: src/understudy/models/ollama.py
"""Ollama provider (local). Native /api/chat for streaming, /api/tags for models."""

from __future__ import annotations

import json
from typing import AsyncIterator

import httpx

from understudy.models.base import ChatMessage, ProviderError


class OllamaProvider:
    kind = "ollama"

    def __init__(
        self,
        *,
        base_url: str = "http://localhost:11434",
        model: str = "llama3.1",
        temperature: float = 0.3,
        timeout: float = 120.0,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.model = model
        self.temperature = temperature
        self._client = httpx.AsyncClient(timeout=timeout)

    async def stream_chat(self, messages: list[ChatMessage]) -> AsyncIterator[str]:
        payload = {
            "model": self.model,
            "messages": messages,
            "stream": True,
            "options": {"temperature": self.temperature},
        }
        try:
            async with self._client.stream("POST", f"{self.base_url}/api/chat", json=payload) as resp:
                if resp.status_code >= 400:
                    body = (await resp.aread()).decode("utf-8", "replace")
                    raise ProviderError(f"HTTP {resp.status_code}: {body[:300]}")
                async for line in resp.aiter_lines():
                    line = line.strip()
                    if not line:
                        continue
                    try:
                        obj = json.loads(line)
                    except json.JSONDecodeError:
                        continue
                    content = (obj.get("message") or {}).get("content")
                    if content:
                        yield content
                    if obj.get("done"):
                        break
        except httpx.HTTPError as exc:
            raise ProviderError(f"cannot reach Ollama at {self.base_url} ({exc})") from exc

    async def list_models(self) -> list[str]:
        try:
            resp = await self._client.get(f"{self.base_url}/api/tags")
        except httpx.HTTPError as exc:
            raise ProviderError(f"cannot reach Ollama at {self.base_url} ({exc})") from exc
        if resp.status_code >= 400:
            raise ProviderError(f"HTTP {resp.status_code}: {resp.text[:200]}")
        return sorted(m.get("name", "") for m in resp.json().get("models", []) if m.get("name"))

    async def check(self) -> str:
        models = await self.list_models()
        if not models:
            return f"Connected @ {self.base_url} · no models — try `ollama pull {self.model}`"
        if self.model not in models:
            return f"Connected · {len(models)} models · '{self.model}' not installed — have: {', '.join(models[:4])}"
        return f"OK · {self.model} @ {self.base_url}"

    async def aclose(self) -> None:
        await self._client.aclose()

File: crates/core/src/models/ollama.rs
//! Ollama provider (local). Native /api/chat streaming, /api/tags for models.

use futures::stream::BoxStream;
use reqwest::Client;
use serde_json::{json, Value};

use super::http::stream_ollama;
use super::{ChatMessage, ProviderError};

pub struct OllamaProvider {
    pub base_url: String,
    pub model: String,
    pub temperature: f32,
    client: Client,
}

impl OllamaProvider {
    pub fn new(base_url: String, model: String, temperature: f32) -> Self {
        OllamaProvider {
            base_url: base_url.trim_end_matches('/').to_string(),
            model,
            temperature,
            client: Client::new(),
        }
    }

    pub fn stream_chat(&self, messages: Vec<ChatMessage>) -> BoxStream<'static, Result<String, ProviderError>> {
        let url = format!("{}/api/chat", self.base_url);
        let payload = json!({
            "model": self.model,
            "messages": messages,
            "stream": true,
            "options": { "temperature": self.temperature },
        });
        Box::pin(stream_ollama(self.client.clone(), url, payload))
    }

    pub async fn list_models(&self) -> Result<Vec<String>, ProviderError> {
        let resp = self
            .client
            .get(format!("{}/api/tags", self.base_url))
            .send()
            .await
            .map_err(|e| ProviderError::msg(format!("cannot reach Ollama at {} ({e})", self.base_url)))?;
        let status = resp.status().as_u16();
        if status >= 400 {
            let body = resp.text().await.unwrap_or_default();
            return Err(ProviderError::msg(format!("HTTP {status}: {}", body.chars().take(200).collect::<String>())));
        }
        let json: Value = resp.json().await.map_err(|e| ProviderError::msg(format!("bad json: {e}")))?;
        let mut names: Vec<String> = json
            .get("models")
            .and_then(|m| m.as_array())
            .map(|a| {
                a.iter()
                    .filter_map(|m| m.get("name").and_then(|n| n.as_str()).map(|s| s.to_string()))
                    .filter(|s| !s.is_empty())
                    .collect()
            })
            .unwrap_or_default();
        names.sort();
        Ok(names)
    }

    pub async fn check(&self) -> Result<String, ProviderError> {
        let models = self.list_models().await?;
        if models.is_empty() {
            return Ok(format!("Connected @ {} · no models — try `ollama pull {}`", self.base_url, self.model));
        }
        if !models.contains(&self.model) {
            let have: Vec<&String> = models.iter().take(4).collect();
            let have = have.iter().map(|s| s.as_str()).collect::<Vec<_>>().join(", ");
            return Ok(format!("Connected · {} models · '{}' not installed — have: {}", models.len(), self.model, have));
        }
        Ok(format!("OK · {} @ {}", self.model, self.base_url))
    }
}
